File: LinkedList/ReverseSLL.py
# Python program to reverse a singly linked list


# Node class
class Node:

    # Constructor to initialise data and next
    def __init__(self, data=None):
        self.data = data
        self.next = None


class SinglyLinkedList:

    # Constructor to initialise head
    def __init__(self):
        self.head = None

    # Function to reverse a linked list
    def reverse(self):

        # If linked list is empty
        if self.head is None:
            return None

        current = self.head
        prev = None

        while current is not None:
            # Store the value of current.next
            next = current.next
            # Set current.next to point to the previous node
            current.next = prev
            # Update pointers for next iteration
            prev = current
            current = next

        self.head = prev

    # Function to Insert data at the beginning of the linked list
    def insert_at_beg(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    # Function to print the linked list
    def print_data(self):
        current = self.head
        while current is not None:
            print(current.data, '-> ', end='')
            current = current.next
        print('None')

if __name__ == '__main__':
    linked_list = SinglyLinkedList()
    linked_list.insert_at_beg(7)
    linked_list.insert_at_beg(6)
    linked_list.insert_at_beg(5)
    linked_list.insert_at_beg(4)
    linked_list.insert_at_beg(3)
    linked_list.insert_at_beg(2)
    linked_list.insert_at_beg(1)
    linked_list.print_data()
    # call the reverse function
    linked_list.reverse()
    # print the reversed list
    linked_list.print_data()

File: LinkedList/ReverseSLL.cpp
// Reverse a singly linked list

#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;

        Node() {
            this->next = NULL;
        }

        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head = NULL;

        void insert_at_beg(int data){
            Node* temp = new Node(data);
            temp->next = this->head;
            this->head = temp;
        }

        void print_data(){
            Node* itr = this->head;

            while(itr != NULL){
                cout << itr->data << " ";
                itr = itr->next;
            }
            cout << endl;
        }
};

void reverse(LinkedList *ll){
    // if linked list is empty or has only 1 node
    if(ll->head == NULL || ll->head->next == NULL) return;

    Node *curr_ptr = ll->head, *next_ptr, *prev_ptr = NULL;

    while(curr_ptr != NULL){
        next_ptr = curr_ptr->next;
        curr_ptr->next = prev_ptr;
        prev_ptr = curr_ptr;
        curr_ptr = next_ptr;
    }

    // update to the new head
    ll->head = prev_ptr;
}

int main() {
    LinkedList* ll = new LinkedList();
    
    // Create Linked list
    ll->insert_at_beg(6);
    ll->insert_at_beg(5);
    ll->insert_at_beg(4);
    ll->insert_at_beg(3);
    ll->insert_at_beg(2);
    ll->insert_at_beg(1);

    cout << "Linked list before reverse: " << endl;
    ll->print_data();

    reverse(ll);
    cout << "Linked list after reverse: " << endl;
    ll->print_data();
    return 0;
}


